### ISSUE TITLE
Proposal for setting and propagating HTTP Header values

### ISSUE THREAD

**Issue by kenfinnigan** [eclipse/microprofile-rest-client#116]
@andymc12 and I discussed this problem at length in today's meeting, which is also referenced as part of #73 and #104. I've created a separate issue to be able to outline some thoughts we came up with today in an effort to garner some additional feedback on the approach.

The spec would define an interface for setting/propagating headers such as:
```
public interface ClientHeaders {
  MultivaluedMap set(MultivaluedMap incomingHeaders, MultivaluedMap clientOutgoingHeaders);
}
```
The purpose of which is to pass in the map of headers from an incoming request and the map of headers that will be set on the outgoing client request based on any annotations that are defined on the service interface, such as `@HeaderParam`. The second parameter, `clientOutgoingHeaders` would be read only in this method. The map returned from this method would then be merged with the map that was passed into `clientOutgoingHeaders` to create the full set of headers for the client request. We pass in the currently defined outgoing headers to provide an opportunity to adjust the values if necessary.

In addition to the interface above, noting that `ClientHeaders` is just a name I came up with not what it will necessarily end up being called, the spec would provide a concrete implementation of the interface that offers the ability to propagate headers from incoming to outgoing based on values within a specific config property. In the short term we may consider adding the headers into a single MP REST Client config property, but longer term it would be better for MP specs to define their own list of headers to propagate that could be read by this concrete class.

Developers then have the ability to define their own classes to perform setting and propagation of headers, but also could extend the spec implementation to include additional header setting that may not be defined by the spec.

So how would this be used?

The spec would define an annotation, possibly called `@ClientHeaderParams`, that can be used on the injection point with `@RESTClient`. By default this annotation would use the concrete implementation from the spec to propagate headers to the client request. If the developer has their own implementation of `ClientHeaders`, they can specify the class on the annotation and that class will be used to propagate/set headers.

In addition, we'd likely add something like `@ClientHeaderParam`, which would take a key and value for setting specific values on the new client request. It would work in a similar fashion to `@HeaderParam`.

We felt it was necessary to offer a way through a method call, `ClientHeaders`, to set a group of header properties as adding a `@ClientHeaderParam` for everything to be set/propagated could be quite lengthy, especially when dealing with tracing headers.

I think I've covered the main pieces we discussed, please correct me @andymc12 if I missed anything.

Please provide feedback on the idea.

**Comment by chunlongliang-ibm** [comment]
The proposed solution does address the token propagation requirement if token exists in request header. However, one of most common scenarios is that token does not exist in the header yet, it is front end web application that need engage 3rd part SSO server to acquire security token (e.g. JWT or oauth access_token, even SAML token),  and client application need propagate the acquired JWT token as Bearer authorization header to call downstream micro services.  Another common scenario is that service client can self-issued and vouches a new token from existing token, and propagate new token to downstream micro service. If the design in this issue is for generic header propagation, it is OK. I would recommend that we need have a different discussion for security token propagation.

**Comment by kenfinnigan** [comment]
@chunlongliang-ibm, the proposal supports what you need by implementing a custom `ClientHeaders` class that can then set whatever it wants onto the outgoing header.

**Comment by chunlongliang-ibm** [comment]
@kenfinnigan , Thanks! Can you send me link to the document and sample to ClientHeaders?

**Comment by kenfinnigan** [comment]
At the moment the only document is the issue description above, which also includes the proposed `ClientHeaders` interface

**Comment by andymc12** [comment]
Per discussion today `@ClientHeaderParams` probably ought to be named something like `@ClientHeaderConverter` (or Generator, Tranferer, etc.) -- if empty, then this tells the MP impl to use the default implementation of `ClientHeaders` - if non-empty, then the user specifies the `ClientHeaders` implementation class to use for converting headers for the outbound client request.

**Comment by andymc12** [comment]
Resolved by PRs #143 and #149.
